File: sc/src/sc/Banque.java
package sc;

import java.io.IOException;

public final class Banque extends BanqueComponent {

   private final IRepository _repository;

   public Banque( String name, IRepository repository ) throws IOException {
      super( name );
      _repository = repository;
      final Thread networkThread = new Thread( this );
      networkThread.setDaemon( true );
      networkThread.start();
   }

   @Override
   public void getInformations( String carteID, SiteCentralGetInformationsResponse response ) throws IOException {
      System.err.printf( getClass().getName() + ".getInformations|carteID = '%s'\n", carteID );
      final ICarte  iCarte  = _repository.getCarte ( carteID );
      final ICompte iCompte = _repository.getCompte( carteID );
      if(( iCarte != null )&&( iCompte != null )) {
         try { Thread.sleep( 3000 ); } catch( final InterruptedException x ) {/**/}
         iCarte .copyTo( response.carte );
         iCompte.copyTo( response.compte );
      }
      else {
         response.carte .id       = "";
         response.carte .code     = "";
         response.carte .month    = 0;
         response.carte .year     = 0;
         response.carte .nbEssais = 3;
         response.compte.id       = "";
         response.compte.solde    = 0.0;
         response.compte.autorise = false;
      }
   }

   @Override
   public void incrNbEssais( String carteID ) {
      System.err.printf( getClass().getName() + ".incrNbEssais( '%s' )\n", carteID );
      final ICarte carte = _repository.getCarte( carteID );
      if( carte != null ) {
         carte.incrNbEssais();
      }
   }

   @Override
   public void retrait( String carteID, double montant ) {
      System.err.printf( getClass().getName() + ".retrait de " + montant + " € à partir de la carte " + carteID + "\n" );
      final ICompte compte = _repository.getCompte( carteID );
      if( compte != null ) {
         compte.retrait( montant );
      }
   }

   @Override
   public void shutdown() {
      _repository.close();
   }

   @Override
   protected void afterDispatch() {
      // Nothing to do
   }
}

File: sc-cpp/src/sc/Banque.cpp
#include <sc/Banque.hpp>

#include <os/Thread.hpp>
#include <os/sleep.hpp>
#include <util/Time.hpp>
#include <util/Log.hpp>

#include <stdio.h>

using namespace sc;

static void * launchUI( void * arg ) {
   BanqueUI * ui = (BanqueUI *)arg;
   ui->run();
   return nullptr;
}

Banque::Banque( const char * name ) :
   BanqueComponent( name ),
   _ui( *this )
{
   UTIL_LOG_MSG( "launching UI" );
   os::Thread( launchUI, &_ui );
}

void Banque::getInformations( const char * carteID, dabtypes::SiteCentralGetInformationsResponse & response ) {
   UTIL_LOG_ARGS( "carteID = %s", carteID );
   dabtypes::Carte * carte = _repository.getCarte( carteID );
   if( carte ) {
      dabtypes::Compte * compte = _repository.getCompte( carteID );
      if( compte ) {
         response.carte  = *carte;
         response.compte = *compte;
      }
      else {
         UTIL_LOG_MSG( "unknown!" );
      }
   }
   else {
      UTIL_LOG_MSG( "unknown!" );
   }
   os::sleep( 3000 );
   UTIL_LOG_DONE();
}

void Banque::incrNbEssais( const char * carteID ) {
   UTIL_LOG_ARGS( "carteID = %s\n", carteID );
   dabtypes::Carte * carte = _repository.getCarte( carteID );
   if( carte ) {
      carte->nbEssais++;
   }
   else {
      UTIL_LOG_MSG( "unknown!" );
   }
   _ui.refresh();
}

void Banque::retrait( const char * carteID, const double & montant ) {
   UTIL_LOG_ARGS( "carteID = %s, montant = %7.2f", carteID, montant );
   dabtypes::Compte * compte = _repository.getCompte( carteID );
   if( compte ) {
      compte->solde -= montant;
   }
   else {
      UTIL_LOG_MSG( "unknown!" );
   }
   _ui.refresh();
}

void Banque::shutdown( void ) {
   UTIL_LOG_HERE();
   terminate();
}
